File: Lab 7/Task2SpecificSensor/app/src/main/java/com/example/task2specificsensor/MainActivity.java
package com.example.task2specificsensor;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity implements SensorEventListener {

    private SensorManager senseMan;
    private Sensor lightSensor;
    private TextView textview;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Map textview to TextView resources
        textview = findViewById(R.id.textview);

        // Map sensor manager to system service
        senseMan = (SensorManager) getSystemService(SENSOR_SERVICE);

        // Map sensor to light sensor
        lightSensor = senseMan.getDefaultSensor(Sensor.TYPE_LIGHT);

        // Register listener for lightSensor
        if (lightSensor != null) {
            senseMan.registerListener(this, lightSensor, SensorManager.SENSOR_DELAY_NORMAL);
            Toast.makeText(this, "Light sensor found", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, "Light sensor not found", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        // Change the value in textview when light level changed
        //float lightLevel = event.values[0];
        textview.setText(String.format("Light Level: %.2f", event.values[0]));
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Implement if needed
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Register listener for lightSensor in onResume
        senseMan.registerListener(this, lightSensor, SensorManager.SENSOR_DELAY_NORMAL);
    }

    @Override
    protected void onPause() {
        super.onPause();
        // Unregister listener for senseMan in onPause
        senseMan.unregisterListener(this);
    }
}

File: Lab 7/Task3Lab7/app/src/main/java/com/example/task3lab7/MainActivity.kt
package com.example.task3lab7

import android.content.Context
import android.hardware.*
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity(), SensorEventListener {

    private lateinit var sensorManager: SensorManager
    private lateinit var accelerometer: Sensor
    private lateinit var gyroscope: Sensor
    private lateinit var magnetometer: Sensor

    private lateinit var accelerometerText: TextView
    private lateinit var gyroscopeText: TextView
    private lateinit var magnetometerText: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        accelerometerText = findViewById(R.id.accelerometerText)
        gyroscopeText = findViewById(R.id.gyroscopeText)
        magnetometerText = findViewById(R.id.magnetometerText)

        sensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager

        accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        gyroscope = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        magnetometer = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)

        registerSensorListeners()
    }

    private fun registerSensorListeners() {
        sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_NORMAL)
        sensorManager.registerListener(this, gyroscope, SensorManager.SENSOR_DELAY_NORMAL)
        sensorManager.registerListener(this, magnetometer, SensorManager.SENSOR_DELAY_NORMAL)
    }

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> updateSensorText(accelerometerText, event.values)
            Sensor.TYPE_GYROSCOPE -> updateSensorText(gyroscopeText, event.values)
            Sensor.TYPE_MAGNETIC_FIELD -> updateSensorText(magnetometerText, event.values)
        }
    }

    private fun updateSensorText(textView: TextView, values: FloatArray) {
        val text = "X=${values[0]}, Y=${values[1]}, Z=${values[2]}"
        runOnUiThread { textView.text = text }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // Do something if accuracy changes
    }

    override fun onDestroy() {
        super.onDestroy()
        sensorManager.unregisterListener(this)
    }
}
